### PR TITLE
4 implement tag handling

### PR DESCRIPTION
**Pull request by ZombieAlienRobot**
## Description of changes
Implemented Tag handling.

  
## How has this been tested?
Please describe the test strategy you followed.

- [X] automated unit test
- [ ] automated integration test
- [ ] automated acceptance test
- [ ] manual, exploratory test

In case of manual test, please document the test well including a set of user instructions and prerequisites. Each including an action, it's result, and where appropriate a screenshot.
In case of automated tests, the test coverage should be reasonably high, especially on Service classes and other classes with complex logic.
    
## Checklist before requesting a review
- [ ] My code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] I have performed a self-review of my code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] My code fulfills all acceptance criteria
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] The test coverage (line and branch) is reasonably high, especially on Service classes and other classes with complex logic
- [ ] I have made corresponding changes to the documentation
- [ ] I have added explanation of architectural decision and rationales to [wiki/adr](https://github.com/IT-REX-Platform/wiki/tree/main/adr)
- [ ] I have updated the changes in the ticket description

## Checklist for reviewer
- [ ] The code works and does not throw errors
- [ ] The code is easy to understand and there are no confusing parts
- [ ] The code follows the [coding guidelines](https://github.com/IT-REX-Platform/wiki/blob/main/dev-manuals/backend/coding-guidelines.md) of this project
- [ ] The code change accomplishes what it is supposed to do
- [ ] I cannot think of any use case in which the code does not behave as intended
- [ ] The added and existing tests reasonably cover the code change
- [ ] I cannot think of any test cases, input or edge cases that should be tested in addition
- [ ] Description of the change is included in the documentation

### CODE REVIEW

File: src/main/java/de/unistuttgart/iste/gits/content_service/persistence/mapper/ContentMapper.java [hunk 1]
```diff
@@ -8,15 +8,21 @@
...
 
     private final ModelMapper modelMapper;
 
-    public ContentDto entityToDto(ContentEntity ContentEntity) {
+    public ContentDto entityToDto(ContentEntity contentEntity) {
         // add specific mapping here if needed
-        return modelMapper.map(ContentEntity, ContentDto.class);
+        ContentDto result = modelMapper.map(contentEntity, ContentDto.class);
+        if (contentEntity.getTags() != null) {
+            result.setTagNames(contentEntity.getTags().stream().map(tag -> tag.getName()).collect(Collectors.toList()));
```
Comment (PaulBredl) [inline src/main/java/de/unistuttgart/iste/gits/content_service/persistence/mapper/ContentMapper.java:23]: You can also just write `.toList()` instead of using `.collect` #Nitpick

File: src/main/java/de/unistuttgart/iste/gits/content_service/service/ContentService.java [hunk 2]
```diff
@@ -21,9 +23,71 @@
...
+        // the newly created tags and the existing tags which are not already assigned to the content need to be linked with the content
+        // for n:m relationships the links have to be created on both sides, i.e. content -> tag and tag --> content
+        tagsToAdd.addAll(tagSynchronizationResult.existingTagsToAdd);
+        if (content.getTags() == null) {
+            content.setTags(tagsToAdd); // TODO clarify if this can really happen or if an empty list is returned
+        } else {
+            content.getTags().addAll(tagsToAdd);
+        }
+        for (TagEntity tag : tagsToAdd) {
+            if (tag.getContents()==null) {
```
Comment (PaulBredl) [inline src/main/java/de/unistuttgart/iste/gits/content_service/service/ContentService.java:71]: formatting

File: src/main/java/de/unistuttgart/iste/gits/content_service/service/ContentService.java [hunk 3]
```diff
@@ -21,9 +23,71 @@
 public class ContentService {
 
     private final ContentRepository contentRepository;
+    private final TagRepository tagRepository;
     private final ContentMapper contentMapper;
     private final ContentValidator contentValidator;
-    
+
+    public static class TagSynchronizationResult {
```
Comment (PaulBredl) [inline src/main/java/de/unistuttgart/iste/gits/content_service/service/ContentService.java:30]: consider extra class

Comment (PaulBredl) [reply src/main/java/de/unistuttgart/iste/gits/content_service/service/ContentService.java:30]: forall this synchronazation stuff

Comment (PaulBredl) [reply src/main/java/de/unistuttgart/iste/gits/content_service/service/ContentService.java:30]: Also records are ideal for such things

File: src/main/java/de/unistuttgart/iste/gits/content_service/persistence/repository/ContentRepository.java [hunk 4]
```diff
@@ -14,11 +14,11 @@ public interface ContentRepository extends JpaRepository<ContentEntity, UUID> {
...
+    @Query("select content from Tag tag join tag.contents content where tag.name = :t")
+    List<ContentEntity> findByTagName(@Param("t") String tag);
 
-    @Query("select t.content from Tag t where t.id = :tagId")
-    List<ContentEntity> findByTagId(@Param("tagId") UUID tagId);
+    @Query("select content from Tag tag join tag.contents content where tag.name in (:tags)")
+    List<ContentEntity> findByTagNames(@Param("tags") List<String> tags);
 
     @Query("select content from Content content where content.id in (:ids)")
     List<ContentEntity> findById(List<UUID> ids);
```
Comment (PaulBredl) [inline src/main/java/de/unistuttgart/iste/gits/content_service/persistence/repository/ContentRepository.java:24]: findAllByIdsIn (naming is a bit confusing)

File: src/test/java/de/unistuttgart/iste/gits/content_service/ContentServiceApiTests.java [hunk 5]
```diff
@@ -13,45 +16,57 @@
 import java.util.List;
 import java.util.UUID;
 
-@GraphQlApiTest
+@SpringBootTest(classes=ContentServiceApplication.class)
+@AutoConfigureGraphQlTester
```
Comment (PaulBredl) [inline src/test/java/de/unistuttgart/iste/gits/content_service/ContentServiceApiTests.java:20]: Inconsistent with other service testing -> no todo in this PR

File: src/test/java/de/unistuttgart/iste/gits/content_service/service/ContentServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+        List<String> tagNames = List.of("Tag1","Tag2","Tag3");
+        TagEntity existingTag = TagEntity.fromName(tagNames.get(1));
+        List<String> newTagNamesToAdd = List.of(tagNames.get(0),tagNames.get(2));
+        ContentService.TagSynchronizationResult result = contentService.prepareSynchronization(tagNames,
+                Collections.emptyList(),
+                List.of(existingTag));
+        assertThat(getTagNames(result.newTagsToAdd), containsInAnyOrder(newTagNamesToAdd.toArray()));
+        assertThat(getTagNames(result.existingTagsToAdd), containsInAnyOrder(List.of(existingTag.getName()).toArray()));
+    }
+    @Test
```
Comment (PaulBredl) [inline src/test/java/de/unistuttgart/iste/gits/content_service/service/ContentServiceTest.java:63]: Formatting

File: src/test/java/de/unistuttgart/iste/gits/content_service/service/ContentServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        assertThat(deletedContentId,is(testContentId));
+        verify(contentRepository).deleteById(testContentId);
+    }
+
+    Set<String> getTagNames(Collection<TagEntity> tags) {
+        return tags.stream().filter(tag -> tag != null).map(tag -> tag.getName()).collect(Collectors.toSet());
+    }
+
+    @Test
+    void testTagSynchronisationNoExistingTagsSuccessful() {
```
Comment (PaulBredl) [inline src/test/java/de/unistuttgart/iste/gits/content_service/service/ContentServiceTest.java:46]: If synchronize is done in separate class it can also be tested in seperate class

File: src/test/java/de/unistuttgart/iste/gits/util/GraphQlApiTest.java [hunk 8]
```diff
@@ -18,7 +18,7 @@
  * <p>
  */
 @ExtendWith(GraphQlTesterParameterResolver.class)
-@ExtendWith(ClearDatabase.class)
+//@ExtendWith(ClearDatabase.class)
```
Comment (PaulBredl) [inline src/test/java/de/unistuttgart/iste/gits/util/GraphQlApiTest.java]: Why??
